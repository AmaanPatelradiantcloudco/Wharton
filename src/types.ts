import { AgentParameters } from "./dialogflow/types";

export type ParticipantToAgentMessage = {
  responseId: string;
  message: string;
  agentId: string;
};

export type AgentToParticipantMessage = {
  responseId: string;
  page: string;
  message: string;
  parameters: AgentParameters;
  done: boolean;
};
