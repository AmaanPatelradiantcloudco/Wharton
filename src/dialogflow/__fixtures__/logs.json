[
  {
    "metadata": {
      "timestamp": "2021-10-21T18:38:19.607825Z",
      "labels": {
        "agent_id": "d232840e-9252-482c-9dc4-3e824d9174af",
        "session_id": "231a7c-f0b-6dd-474-b15bfdcf4"
      }
    },
    "data": {
      "responseId": "49b7672d-62a8-4b98-8529-36f45289ae8d",
      "queryResult": {
        "text": "No",
        "intent": {
          "displayName": "If they ask why I asked about robot",
          "name": "projects/wharton-chatbot-test/locations/us-central1/agents/d232840e-9252-482c-9dc4-3e824d9174af/intents/2a608514-d697-4092-a28c-a5315f5d2804"
        },
        "currentPage": {
          "displayName": "Information Exchange",
          "name": "projects/wharton-chatbot-test/locations/us-central1/agents/d232840e-9252-482c-9dc4-3e824d9174af/flows/00000000-0000-0000-0000-000000000000/pages/1884e9b2-2a95-4fd2-a564-414b69d962b7"
        },
        "intentDetectionConfidence": 0.51311874,
        "responseMessages": [
          {
            "text": {
              "text": [
                "Just wanted to check. I’m interested in buying the Hearts painting. Can you tell me more about it?"
              ]
            },
            "responseType": "HANDLER_PROMPT",
            "source": "VIRTUAL_AGENT"
          }
        ],
        "match": {
          "intent": {
            "displayName": "If they ask why I asked about robot",
            "name": "projects/wharton-chatbot-test/locations/us-central1/agents/d232840e-9252-482c-9dc4-3e824d9174af/intents/2a608514-d697-4092-a28c-a5315f5d2804"
          },
          "modelType": "MODEL_TYPE_STANDARD",
          "resolvedInput": "No",
          "confidence": 0.51311874,
          "matchType": "INTENT"
        }
      },
      "responseType": "FINAL"
    }
  },
  {
    "metadata": {
      "timestamp": "2021-10-21T18:38:24.851268Z",
      "labels": {
        "session_id": "231a7c-f0b-6dd-474-b15bfdcf4",
        "agent_id": "d232840e-9252-482c-9dc4-3e824d9174af"
      }
    },
    "data": {
      "responseType": "FINAL",
      "responseId": "3bc8b6cb-28f3-4f87-8f95-e7af744f9b48",
      "queryResult": {
        "responseMessages": [
          {
            "source": "VIRTUAL_AGENT",
            "text": {
              "text": ["What is your initial offer?"]
            },
            "responseType": "PARAMETER_PROMPT"
          }
        ],
        "text": "It's cool",
        "currentPage": {
          "displayName": "Bargaining",
          "name": "projects/wharton-chatbot-test/locations/us-central1/agents/d232840e-9252-482c-9dc4-3e824d9174af/flows/00000000-0000-0000-0000-000000000000/pages/9d6e62bc-252b-4c97-9631-cd1086277a07"
        },
        "match": {
          "confidence": 0.3,
          "modelType": "MODEL_TYPE_STANDARD",
          "matchType": "NO_MATCH"
        },
        "intentDetectionConfidence": 0.3
      }
    }
  },
  {
    "metadata": {
      "timestamp": "2021-10-21T18:38:28.267190Z",
      "labels": {
        "agent_id": "d232840e-9252-482c-9dc4-3e824d9174af",
        "session_id": "231a7c-f0b-6dd-474-b15bfdcf4"
      }
    },
    "data": {
      "responseType": "FINAL",
      "queryResult": {
        "currentPage": {
          "name": "projects/wharton-chatbot-test/locations/us-central1/agents/d232840e-9252-482c-9dc4-3e824d9174af/flows/00000000-0000-0000-0000-000000000000/pages/9d6e62bc-252b-4c97-9631-cd1086277a07",
          "displayName": "Bargaining"
        },
        "intentDetectionConfidence": 1,
        "parameters": {
          "offer1": "10000.0"
        },
        "text": "10000",
        "responseMessages": [
          {
            "text": {
              "text": [
                "I’m afraid that is more than I expected to pay. Could you do $7,000? "
              ]
            },
            "source": "VIRTUAL_AGENT",
            "responseType": "HANDLER_PROMPT"
          }
        ],
        "match": {
          "parameters": {
            "offer1": "10000.0"
          },
          "modelType": "MODEL_TYPE_STANDARD",
          "matchType": "PARAMETER_FILLING",
          "resolvedInput": "10000",
          "confidence": 1
        }
      },
      "responseId": "5d4a2b9d-6990-4b32-9a92-158d5e045e2e"
    }
  }
]
