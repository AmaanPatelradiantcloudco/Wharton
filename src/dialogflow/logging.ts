import { Logging } from "@google-cloud/logging";
import { transformMessages } from "./helpers";
import { LogParameter, LogResponse } from "./types";

export async function* readResponses(
  projectId: string,
  month: Date,
  logging = new Logging({ projectId })
): AsyncIterable<LogResponse> {
  const start = new Date(month.getFullYear(), month.getMonth(), 1);
  const end = new Date(month.getFullYear(), month.getMonth() + 1, 0);

  const stream = logging.getEntriesStream({
    maxApiCalls: 20,
    autoPaginate: true,
    filter: `
      logName = "projects/${projectId}/logs/dialogflow-runtime.googleapis.com%2Frequests"
      timestamp >= "${start.toISOString()}"
      timestamp <= "${end.toISOString()}"
      jsonPayload.responseId=~".+"
    `,
  });

  for await (const { data, metadata } of stream) {
    yield {
      responseId: data.responseId ?? "",
      agentId: metadata.labels?.agent_id ?? "",
      sessionId: metadata.labels?.session_id ?? "",
      timestamp: metadata.timestamp?.toString() ?? "",
      request: data.queryResult?.text ?? "",
      response: transformMessages(data.queryResult?.responseMessages ?? []),
      parameters: data.queryResult?.parameters ?? {},
    };
  }
}

export const expandParameters = (response: LogResponse): LogParameter[] => {
  const { parameters, ...info } = response;

  return Object.entries(parameters).map(([name, value]) => ({
    ...info,
    name,
    value,
  }));
};
