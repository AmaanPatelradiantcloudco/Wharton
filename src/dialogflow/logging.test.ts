import { Readable } from "stream";
import { expandParameters, readResponses } from "./logging";
import { LogResponse } from "./types";
import logs from "./__fixtures__/logs.json";

const createLogging = () => ({
  getEntriesStream: jest.fn(() => {
    const stream = new Readable({ objectMode: true });
    logs.forEach((log) => stream.push(log));
    stream.push(null);
    return stream;
  }),
});

test("readResponses", async () => {
  const month = new Date();
  const logging = createLogging();
  const stream = readResponses("project id", month, logging as any);
  const responses: LogResponse[] = [];

  for await (const response of stream) {
    responses.push(response);
  }

  expect(responses).toHaveLength(3);
  expect(responses).toMatchSnapshot();
});

test("expandParameters", () => {
  const shared = {
    agentId: "agentId",
    request: "request",
    response: "response",
    responseId: "responseId",
    sessionId: "sessionId",
    timestamp: "timestamp",
  };

  const response: LogResponse = {
    ...shared,
    parameters: { a: 1, b: 2 },
  };

  expect(expandParameters(response)).toEqual([
    { ...shared, name: "a", value: 1 },
    { ...shared, name: "b", value: 2 },
  ]);
});
