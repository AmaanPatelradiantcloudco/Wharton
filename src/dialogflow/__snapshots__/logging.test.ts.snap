// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`readResponses 1`] = `
Array [
  Object {
    "agentId": "d232840e-9252-482c-9dc4-3e824d9174af",
    "parameters": Object {},
    "request": "No",
    "response": "Just wanted to check. I’m interested in buying the Hearts painting. Can you tell me more about it?",
    "responseId": "49b7672d-62a8-4b98-8529-36f45289ae8d",
    "sessionId": "231a7c-f0b-6dd-474-b15bfdcf4",
    "timestamp": "2021-10-21T18:38:19.607825Z",
  },
  Object {
    "agentId": "d232840e-9252-482c-9dc4-3e824d9174af",
    "parameters": Object {},
    "request": "It's cool",
    "response": "What is your initial offer?",
    "responseId": "3bc8b6cb-28f3-4f87-8f95-e7af744f9b48",
    "sessionId": "231a7c-f0b-6dd-474-b15bfdcf4",
    "timestamp": "2021-10-21T18:38:24.851268Z",
  },
  Object {
    "agentId": "d232840e-9252-482c-9dc4-3e824d9174af",
    "parameters": Object {
      "offer1": "10000.0",
    },
    "request": "10000",
    "response": "I’m afraid that is more than I expected to pay. Could you do $7,000? ",
    "responseId": "5d4a2b9d-6990-4b32-9a92-158d5e045e2e",
    "sessionId": "231a7c-f0b-6dd-474-b15bfdcf4",
    "timestamp": "2021-10-21T18:38:28.267190Z",
  },
]
`;
